### PR TITLE
[FLINK-25044][testing][Pulsar Connector] Add More Unit Test For Pulsar Source

### PR DESCRIPTION
**Pull request by imaffe**
## What is the purpose of the change

Enhance the pulsar source unit test for the following classes
- PulsarSourceReader
- PulsarPartitionSplitReader
- PulsarSourceEnumerator
- PulsarSourceBuilder

## Brief change log

One bug fix with this test enhancement:
1. [BUG FIX] Fix overflow when stopCursor is latest and use exclusive start cursor .
2. Added tests for Ordered and Unordered version of PulsarSourceReader and PulsarPartitionSplitReader
3. Improved tests for PulsarSourceEnumerator and PulsarSourceBuilder

## Verifying this change

Local Test failed due to a test failure not related to this change, see if that can pass in the CI.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? ( no)

### CODE REVIEW

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarOrderedSourceReaderTest.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        config.set(PulsarSourceOptions.PULSAR_SUBSCRIPTION_TYPE, SubscriptionType.Failover);
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    void pollMessageWithIntervalLongerThanTimeout(
+            boolean autoAcknowledgementEnabled, TestInfo testInfo) throws Exception {
+        PulsarSourceReaderBase<Integer> reader = sourceReader(autoAcknowledgementEnabled);
+        Boundedness boundedness = Boundedness.CONTINUOUS_UNBOUNDED;
+        String topicName = uniqueTestName(testInfo);
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarOrderedSourceReaderTest.java:62]: Isn't this topic created before every test in `PulsarSourceReaderTestBase#before`?

In general, I'd discourage sharing test setup code through inheritance because it becomes unmaintainable very fast and it is hard to extend. I prefer to use extensions that I can use in any test.

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+        reader.addSplits(Collections.singletonList(emptySplit));
+
+        TestingReaderOutput<Integer> output = new TestingReaderOutput<>();
+        InputStatus status;
+        status = reader.pollNext(output);
+        // We sleep for 1 sec to avoid entering false negative case:
+        // we perform assertion when add splits task
+        // is dequeued but not executed.
+        sleepUninterruptibly(1, TimeUnit.SECONDS);
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java:124]: Can you explain why we need this? This also seems like a source of flakiness.

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+        assertEquals(status, InputStatus.NOTHING_AVAILABLE);
+        reader.close();
+    }
+
+    // ---------------------
+    protected String uniqueTestName(TestInfo testInfo) {
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java:130]: Please move all util methods into either a dedicated helper class if they can be static or introduce a new test extension for pulsar

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarOrderedPartitionSplitReaderTest.java [hunk 4]
```diff
@@ -34,4 +43,57 @@
...
+    }
+
+    @Test
+    void consumeMessageCreatedBeforeHandleSplitsChangesAndUseEarliestStartCursorWithoutSeek() {
+        PulsarPartitionSplitReaderBase<String> splitReader = splitReader();
+        String topicName = randomAlphabetic(10);
+        operator().setupTopic(topicName, STRING, () -> randomAlphabetic(10));
+        handleSplit(splitReader, topicName, 0, MessageId.earliest);
+        fetchedMessages(splitReader, NUM_RECORDS_PER_PARTITION, true);
+    }
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarOrderedPartitionSplitReaderTest.java:72]: Nit: Maybe think about using a parameterized test to avoid the code duplication

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java [hunk 5]
```diff
@@ -18,95 +18,165 @@
...
+        return fetchedMessages(splitReader, 1, false).stream().findFirst().orElse(null);
+    }
+
+    protected <T> List<PulsarMessage<T>> fetchedMessages(
+            PulsarPartitionSplitReaderBase<T> splitReader, int expectedCount, boolean verify) {
+        return fetchedMessages(
+                splitReader, expectedCount, verify, Boundedness.CONTINUOUS_UNBOUNDED);
+    }
+
+    protected <T> List<PulsarMessage<T>> fetchedMessages(
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:136]: Please move these methods to a util class or test extension

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java [hunk 6]
```diff
@@ -126,63 +196,116 @@ void pollMessageAfterTimeout(PulsarPartitionSplitReaderBase<String> splitReader)
...
-                    });
-        }
+    @Test
+    void assignNoSplits() {
+        PulsarPartitionSplitReaderBase<String> splitReader = splitReader();
+        assertNull(fetchedMessage(splitReader));
     }
+
+    /** Create a split reader with max message 1, fetch timeout 1s. */
+    protected abstract PulsarPartitionSplitReaderBase<String> splitReader();
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:310]: You can implement an extension to instantiate the test SplitReader in the subclasses and directly passes it to the test without the need of an abstract method. You can take a look at how the `SourceTestSuiteBase` does it with the `TestEnvironment` [1]

[1] https://github.com/apache/flink/blob/master/flink-test-utils-parent/flink-connector-testing/src/main/java/org/apache/flink/connectors/test/common/testsuites/SourceTestSuiteBase.java

**Review by fapaul — CHANGES_REQUESTED** [review]
Thanks for the additional tests it will definitely help to increase the stability of the connector 🙏 

I left a few remarks regarding the test setup and had some questions.
Can you also try to split the PR in separate commits?

1. Bug fix code + test
2. More tests (Maybe split again to ease review i.e. one commit for splitreader one for sourcereader)

**Comment by imaffe** [comment]
> Thanks for the additional tests it will definitely help to increase the stability of the connector 🙏
> 
> I left a few remarks regarding the test setup and had some questions. Can you also try to split the PR in separate commits?
> 
> 1. Bug fix code + test
> 2. More tests (Maybe split again to ease review i.e. one commit for splitreader one for sourcereader)

Great, I'll be working on it.

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumeratorTest.java [hunk 7]
```diff
@@ -112,64 +119,270 @@ void startWithPeriodicPartitionDiscovery() throws Exception {
...
+            runOneTimePartitionDiscovery(context);
+            verifyLastReadersAssignments(subscriptionType, context, prexistingTopics, 1);
+        }
+    }
+
+    @ParameterizedTest
+    @EnumSource(
+            value = SubscriptionType.class,
+            names = {"Failover", "Shared"})
+    @Timeout(value = 60)
```
[the remark this answers is not in the document]
Comment (imaffe) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumeratorTest.java:154]: Yeah I recall there is a azure timeout for each test. If that is the case I don't need the timeout on JUnit.

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarOrderedSourceReaderTest.java [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+                        < NUM_RECORDS_PER_PARTITION * DEFAULT_PARTITIONS);
+
+        // The completion of the last checkpoint should subsume all previous checkpoints.
+        assertEquals(checkpointId, reader.cursorsToCommit.size());
+
+        long lastCheckpointId = checkpointId;
+        CommonTestUtils.waitUtil(
+                () -> {
+                    try {
+                        reader.notifyCheckpointComplete(lastCheckpointId);
```
[the remark this answers is not in the document]
Comment (imaffe) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarOrderedSourceReaderTest.java:148]: I think the logic is when the checkpoint complete and the  cursorsToCommit is empty, then it means the checkpoint is OK. It will run multiple times if there is an exception. Ideally after one run the cursorsToCommit should become empty. But I do agree maybe we can optimize this so that if one exception is thrown, we consider the test as fail.

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java (hunk 2 above)
Comment (imaffe) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java:124]: Oh, there is an edge case. Basically the fetcher could become idle upon entry because : 1. the AddSplitsTask is dequeued from taskQueue, 2. The dequeued splits is not assigned to `assignedSplits`.And if both `assignedSplits` is null and taskQueue is empty, the fetcher could become idle, and return NOTHING_AVAILABLE upon entry. This happened multiple times during local test.

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarOrderedPartitionSplitReaderTest.java (hunk 4 above)
Comment (imaffe) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarOrderedPartitionSplitReaderTest.java:72]: Oh they are different test cases. Using latest and latest - 1 has different semantics (and different code paths). But sure Parameterized test will also be feasible.

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java (hunk 6 above)
Comment (imaffe) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:310]: Do you mean use @TestTemplate and a custom TestInvocationContextProvider ? Me and @syhily used @TestTemplate and TestInvocationContext and then refactored to what is looks like now (cause we thought they were an overkill in our case).  But sure I can switch back to use the TestInvocationContextProvider for extensibility.

**Comment by imaffe** [comment]
@flinkbot run azure

**Comment by imaffe** [comment]
@flinkbot run azure

**Comment by imaffe** [comment]
@fapaul Hi Fabian, I've updated this PR, would you mind take a look ? Thanks ~

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+                                "Caught unexpected exception when polling from the reader",
+                                exception);
+                    }
+                    return condition.get();
+                },
+                Duration.ofSeconds(DEFAULT_POLLING_TIMEOUT_SECONDS),
+                errorMessage);
+    }
+
+    protected void verifyAllMessageAcknowledged(int expectedMessages, String partitionName)
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java:193]: I wonder whether we can introduce an assertion helper class that is constructed with the `operator`. So we do not need to use inheritance to share utils

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.fail;
+
+@ExtendWith({
+    PulsarSourceOrderlinessExtension.class,
+    TestLoggerExtension.class,
+})
+abstract class PulsarSourceReaderTestBase extends PulsarTestSuiteBase {
+
+    protected static final int DEFAULT_POLLING_TIMEOUT_SECONDS = 120;
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java:92]: private

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarOrderedSourceReaderTest.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+    private static final int MAX_EMPTY_POLLING_TIMES = 10;
+
+    @SubType SubscriptionType subscriptionType = SubscriptionType.Failover;
+
+    @TestTemplate
+    void pollMessageWithIntervalLongerThanTimeout(
+            PulsarSourceReaderBase<Integer> reader, Boundedness boundedness, String topicName)
+            throws Exception {
+        setupSourceReader(reader, topicName, 0, boundedness);
+        sleepUninterruptibly(15, TimeUnit.SECONDS);
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarOrderedSourceReaderTest.java:51]: Can you make the test faster than 15 seconds?

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java [hunk 12]
```diff
@@ -39,78 +50,164 @@
...
+    TestLoggerExtension.class,
+})
 public abstract class PulsarPartitionSplitReaderTestBase extends PulsarTestSuiteBase {
 
     @RegisterExtension
     PulsarSplitReaderInvocationContextProvider provider =
             new PulsarSplitReaderInvocationContextProvider();
 
+    /** Default reader config: max message 1, fetch timeout 1s. */
     protected Configuration readerConfig() {
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:96]: private?

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java [hunk 13]
```diff
@@ -39,78 +50,164 @@
...
         Configuration config = operator().config();
         config.set(PULSAR_MAX_FETCH_RECORDS, 1);
         config.set(PULSAR_MAX_FETCH_TIME, 1000L);
         config.set(PULSAR_SUBSCRIPTION_NAME, randomAlphabetic(10));
         config.set(PULSAR_ENABLE_AUTO_ACKNOWLEDGE_MESSAGE, true);
-
         return config;
     }
 
     protected SourceConfiguration sourceConfig() {
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:105]: private?

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java [hunk 14]
```diff
@@ -39,78 +50,164 @@
...
+        SplitsAddition<PulsarPartitionSplit> addition = new SplitsAddition<>(singletonList(split));
+        reader.handleSplitsChanges(addition);
+    }
+
+    protected void seekStartPositionAndHandleSplit(
+            PulsarPartitionSplitReaderBase<String> reader, String topicName, int partitionId) {
+        seekStartPositionAndHandleSplit(reader, topicName, partitionId, MessageId.latest);
+    }
+
+    protected void seekStartPositionAndHandleSplit(
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:131]: private?

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java [hunk 15]
```diff
@@ -39,78 +50,164 @@
...
+            int partitionId,
+            MessageId startPosition) {
+        TopicPartition partition = new TopicPartition(topicName, partitionId, createFullRange());
+        PulsarPartitionSplit split =
+                new PulsarPartitionSplit(partition, StopCursor.never(), startPosition, null);
+        SplitsAddition<PulsarPartitionSplit> addition = new SplitsAddition<>(singletonList(split));
+        reader.handleSplitsChanges(addition);
+    }
+
+    protected void seekStartPositionAndHandleSplit(
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:126]: private?

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java [hunk 16]
```diff
@@ -39,78 +50,164 @@
...
+            StartCursor startCursor = StartCursor.fromMessageId(startPosition);
+            startCursor.seekPosition(partition.getTopic(), partition.getPartitionId(), consumer);
+        } catch (PulsarClientException e) {
+            sneakyThrow(e);
+        }
+
+        reader.handleSplitsChanges(addition);
     }
 
     protected <T> PulsarMessage<T> fetchedMessage(PulsarPartitionSplitReaderBase<T> splitReader) {
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:153]: private?

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java [hunk 17]
```diff
@@ -39,78 +50,164 @@
...
+        return fetchedMessages(splitReader, 1, false).stream().findFirst().orElse(null);
+    }
+
+    protected <T> List<PulsarMessage<T>> fetchedMessages(
+            PulsarPartitionSplitReaderBase<T> splitReader, int expectedCount, boolean verify) {
+        return fetchedMessages(
+                splitReader, expectedCount, verify, Boundedness.CONTINUOUS_UNBOUNDED);
+    }
+
+    protected <T> List<PulsarMessage<T>> fetchedMessages(
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:163]: private?

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java [hunk 18]
```diff
@@ -39,78 +50,164 @@
...
 import static org.apache.pulsar.client.api.Schema.STRING;
-import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.assertj.core.api.Assertions.assertThat;
 
 /** Test utils for split readers. */
+@ExtendWith({
+    PulsarSourceOrderlinessExtension.class,
+    TestLoggerExtension.class,
+})
 public abstract class PulsarPartitionSplitReaderTestBase extends PulsarTestSuiteBase {
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:89]: A lot of the methods of this class seem to unnecessarily be protected either they can be private or even better provide util outside if they do not work with an instance reference.

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/testutils/extension/PulsarSourceOrderlinessExtension.java [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.platform.commons.support.AnnotationSupport;
+
+import java.lang.annotation.Annotation;
+import java.util.Collection;
+import java.util.List;
+
+/** An extension for subclasses to specify {@link org.apache.pulsar.client.api.SubscriptionType}. */
+public class PulsarSourceOrderlinessExtension implements BeforeAllCallback {
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/testutils/extension/PulsarSourceOrderlinessExtension.java:31]: Is orderliness the same as ordered here?

File: flink-test-utils-parent/flink-test-utils-junit/pom.xml [hunk 20]
```diff
@@ -21,6 +21,10 @@ under the License.
 	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
 
 	<modelVersion>4.0.0</modelVersion>
+	<properties>
+		<maven.compiler.source>8</maven.compiler.source>
+		<maven.compiler.target>8</maven.compiler.target>
+	</properties>
```
Comment (fapaul) [inline flink-test-utils-parent/flink-test-utils-junit/pom.xml:27]: Why is this needed? I am not sure whether we want to do this because usually all modules should work with java 8 and java 11

**Review by fapaul — COMMENTED** [review]
Thanks for the update. It already looks much better I left some cleanup comments and one comment regarding a potential unstableness.

File: flink-test-utils-parent/flink-test-utils-junit/pom.xml (hunk 20 above)
Comment (imaffe) [reply flink-test-utils-parent/flink-test-utils-junit/pom.xml:27]: This should be deleted (I added this line for local debugging purposes), sorry for importing this one.

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/testutils/extension/PulsarSourceOrderlinessExtension.java (hunk 19 above)
Comment (imaffe) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/testutils/extension/PulsarSourceOrderlinessExtension.java:31]: The word choice is a bit weird, but what I want to express here is this extension passes either ordered or unordered as a parameter to the tes. I'll try to rephrase it

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java (hunk 18 above)
Comment (imaffe) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:89]: Many of the methods might be used by the subclaseses. I'll look at these methods to see which can be private or put them in util classes

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java [hunk 21]
```diff
@@ -0,0 +1,323 @@
...
+        // 1. the AddSplitsTask is dequeued from taskQueue,
+        // 2. The dequeued splits is not assigned to assignedSplits.
+        // And if both assignedSplits is null and taskQueue is empty, the fetcher could become idle,
+        // and return NOTHING_AVAILABLE upon entry. This happened multiple times during local test.
+        sleepUninterruptibly(1, TimeUnit.SECONDS);
+        assertThat(status).isEqualTo(InputStatus.NOTHING_AVAILABLE);
+        reader.close();
+    }
+
+    protected void setupSourceReader(
```
[the remark this answers is not in the document]
Comment (imaffe) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java:146]: I'll try to avoid using this sleep and see if we have other choices

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java [hunk 22]
```diff
@@ -0,0 +1,323 @@
...
+            throws Exception {
+        final PulsarPartitionSplit emptySplit =
+                createPartitionSplit(
+                        topicName, 0, Boundedness.CONTINUOUS_UNBOUNDED, MessageId.latest);
+
+        reader.addSplits(Collections.singletonList(emptySplit));
+
+        TestingReaderOutput<Integer> output = new TestingReaderOutput<>();
+        InputStatus status;
+        status = reader.pollNext(output);
```
[the remark this answers is not in the document]
Comment (imaffe) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java:134]: Basically the new test cases covers this feature change. Without this fix, one of the newly added test will fail.

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java (hunk 17 above)
Comment (imaffe) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:163]: I think here these methodsare provided potentially for the subclasses.

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java (hunk 18 above)
Comment (imaffe) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:89]: For the general design guideline, do we prefer not using protected methods. For example, in design method X is designed to be exposed to its subclasses, but currenly subclass is not using that method. Do we keep method X as protected, or we change X to private, and only change X to protected again when we needed to use that method in the subclasses ?

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (imaffe) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java:146]: This sleep can be deleted here as this comment is not relevant. The relevant test case was deleted because that test case is covered in other test cases

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java (hunk 18 above)
Comment (fapaul) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:89]: In general, I think having less visibility is preferable so using private where ever possible. Furthermore, there is also a saying that one should use composition over inheritance. Every time I want to use inheritance I think for a moment if there is no other way before implementing it.

Comment (imaffe) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:89]: I used private and moved methods to subclasses.

Comment (imaffe) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:89]: For composition over inheritance, I totally understand the design considerations, but I kinda feel currently using composition here is a little bit overkill from readability perspective. I think once the testing methods continues to grow (if needed) we can consider using a class to place those methods later.

Comment (imaffe) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java:89]: And this new fix should have covered the comments ~ Let me know if we need other fixes or I missed anything ~ Thank you so much @fapaul for valuable feedbacks ~

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarOrderedSourceReaderTest.java (hunk 11 above)
Comment (fapaul) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarOrderedSourceReaderTest.java:63]: @imaffe can you have a look at this?

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java (hunk 10 above)
Comment (fapaul) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java:84]: seems not addressed

**Review by fapaul — APPROVED** [review]
I think two comments are not addressed but everything else looks good now, thanks for the patience. Can you squash all the commits together that the PR only includes two commits one for the bug fix and one for the tests?
Nit: Next time it would be great if you mark the comments you have addressed :)

**Comment by imaffe** [comment]
> I think two comments are not addressed but everything else looks good now, thanks for the patience. Can you squash all the commits together that the PR only includes two commits one for the bug fix and one for the tests? Nit: Next time it would be great if you mark the comments you have addressed :)

The first case can be deleted. The edge case it tries to test is addressed in the SplitReaderTest. Basically what we want to do here is to wait for a certain timeout between two messages in the pulsar client and see how it behaves. Do we have suggestions on how to test these kinds of behaviour? (needs to wait for an external system to timeout, feels we can't use a blocking queue to signal a event complete here). 

The second place, I'll set the polling timeout to a large number. And rely on the global timeout to terminate execution if the end condition is not met.

**Comment by imaffe** [comment]
It includes 3 commits, if the newest is ok, I'll squash to 2 commits ~

**Comment by fapaul** [comment]
@flinkbot run azure
